sdk/core/core-rest-pipeline/src/policies/auxiliaryAuthenticationHeaderPolicy.ts

```diff
@@ -0,0 +1,106 @@
+const AUTHORIZATION_AUXILIARY_HEADER = "x-ms-authorization-auxiliary";
+
+/**
+ * Options to configure the auxiliaryAuthenticationHeaderPolicy
+ */
+export interface AuxiliaryAuthenticationHeaderPolicyOptions {
+  /**
+   * TokenCredential list used to get token from auxiliary tenants and
+   * one credential for each tenant the client may need to access
+   */
+  credentials?: TokenCredential[];
+  /**
+   * Scopes depend on the cloud your application runs in
+   */
+  scopes: string | string[];
+  /**
+   * A logger can be sent for debugging purposes.
+   */
+  logger?: AzureLogger;
+}
+
+type NullableString = string | null | undefined;
+
+async function sendAuthorizeRequest(options: AuthorizeRequestOptions): Promise<NullableString> {
+  const { scopes, getAccessToken, request } = options;
+  const getTokenOptions: GetTokenOptions = {
+    abortSignal: request.abortSignal,
+    tracingOptions: request.tracingOptions,
+  };
+
+  return (await getAccessToken(scopes, getTokenOptions))?.token;
+}
+
+/**
+ * A policy for external tokens to `x-ms-authorization-auxiliary` header.
+ * This header will be used when creating a cross-tenant application we may need to handle authentication requests
+ * for resources that are in different tenants.
+ * You could see [ARM docs](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant) for a rundown of how this feature works
+ */
+export function auxiliaryAuthenticationHeaderPolicy(
+  options: AuxiliaryAuthenticationHeaderPolicyOptions
+): PipelinePolicy {
+  const { credentials, scopes } = options;
+  const logger = options.logger || coreLogger;
+  const tokenCyclerMap = new WeakMap<TokenCredential, AccessTokenGetter>();
+
+  return {
+    name: auxiliaryAuthenticationHeaderPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      if (!request.url.toLowerCase().startsWith("https://")) {
+        throw new Error(
+          "Bearer token authentication for auxiliary header is not permitted for non-TLS protected (non-https) URLs."
+        );
+      }
+      if (!credentials || credentials.length === 0) {
+        logger.info(`Skip ${auxiliaryAuthenticationHeaderPolicyName} due to empty credentials`);
+        return next(request);
+      }
+
+      const tokenPromises: Promise<NullableString>[] = [];
+      for (const credential of credentials) {
+        if (!tokenCyclerMap.has(credential)) {
+          tokenCyclerMap.set(credential, createTokenCycler(credential));
+        }
+        tokenPromises.push(
+          sendAuthorizeRequest({
+            scopes: Array.isArray(scopes) ? scopes : [scopes],
+            request,
+            getAccessToken: tokenCyclerMap.get(credential)!,
+            logger,
+          })
+        );
+      }
+      const auxiliaryTokens: NullableString[] = (await Promise.all(tokenPromises)).filter((token) =>
```

Comment:
`auxiliaryTokens` is actually `string[]` since we are filtering out null/undefined. To help TS realize this we can use an assertion on filter's callback:

```suggestion
      const auxiliaryTokens = (await Promise.all(tokenPromises)).filter((token): token is string =>
```